### PR TITLE
[WIP]support openpgp

### PR DESCRIPTION
**Pull request by la-luo**
## Description of changes in release / Impact of release:
(insert text here)

## Documentation
(insert text here)

### Is this a breaking change?
- [ ] Yes
- [ ] No

### If you answered Yes then describe why is it so
(insert text here if applicable)

### Is there a way to disable the change?
- [ ] Use previous release
- [ ] Use a feature flag
- [ ] No

#### Additional details go here
(insert text here if applicable)

### CODE REVIEW

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star [hunk 1]
```diff
@@ -0,0 +1,265 @@
+load("@stdlib//larky", WHILE_LOOP_EMULATION_ITERATION="WHILE_LOOP_EMULATION_ITERATION", larky="larky")
+load("@stdlib//struct", struct="struct")
+load("@vendor//Crypto/Util/py3compat", tobytes="tobytes", bord="bord", tostr="tostr")
+load("@vendor//option/result", Error="Error")
+
+StopIterating = Error("StopIteration")
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:6]: This is now StopIteration. You can see the test_builtins.star on how to use it. https://github.com/verygoodsecurity/starlarky/blob/e72ace569282a7cd6dd36aace70fb9178e499cda/larky/src/test/resources/stdlib_tests/test_builtins.star#L63

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+def Packet():
+    """ OpenPGP packet.
+        http://tools.ietf.org/html/rfc4880#section-4.1
+        http://tools.ietf.org/html/rfc4880#section-4.3
+    """
+
+    self = larky.mutablestruct(__name__='Packet', __class__=Packet)
+
+    def __init__(data=None):
+        for tag in Packet.tags:
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:102]: What is Packet.tags? Did you mean self.tags?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+    def parse(input_data):
+        if hasattr(input_data, 'next') or hasattr(input_data, '__next__'):
+            g = PushbackGenerator(input_data)
+        else:
+            g = PushbackGenerator(_gen_one(input_data))
+
+        packet = None
+        # If there is not even one byte, then there is no packet at all
+        # chunk = _ensure_bytes(1, next(g), g)
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:117]: FYI - next() and iter() now work in Larky. https://github.com/verygoodsecurity/starlarky/pull/161

https://github.com/verygoodsecurity/starlarky/blob/ceb142820d497baeabd9fde4b58001212904e280/larky/src/test/resources/stdlib_tests/test_builtins.star#L167

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    return self
+
+def LiteralDataPacket(Packet):
+    """ OpenPGP Literal Data packet (tag 11).
+        http://tools.ietf.org/html/rfc4880#section-5.9
+    """
+    self = larky.mutablestruct(__name__='LiteralDataPacket', __class__=LiteralDataPacket)
+
+    def __init__(data=None, format='b', filename='data', timestamp=time()):
+        super(LiteralDataPacket, self).__init__()
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:232]: Super does not work in Larky. Take a look at the csv module to see how I was able to do "inheritance".

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        if hasattr(data, 'encode'):
+            data = data.encode('utf-8')
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:234]: codecs.encode?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        self.data = data
+        self.format = format
+        self.filename = filename.encode('utf-8')
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:237]: codecs.encode?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+        der = [b'\x30']
+        ulen = len(r) + len(s) + 4
+        if ulen >= 128:
+            der += [b'\x81']
+        der += [pack('!B', ulen)]
+        der += [b'\x02', pack('!B', len(r)), r]
+        der += [b'\x02', pack('!B', len(s)), s]
+        return b''.join(der)
+    self._encode_dsa_der = _encode_dsa_der
+    _encode_dsa_der = _encode_dsa_der
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:560]: Needed?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+    self._bytes_to_long = _bytes_to_long
+    _bytes_to_long = _bytes_to_long
+
+    def _block_pad_unpad(cls, siz, bs, go):
+        pad_amount = siz - (len(bs) % siz)
+        return go(bs + b'\0'*pad_amount)[:-pad_amount]
+    self._block_pad_unpad = _block_pad_unpad
+    _block_pad_unpad = _block_pad_unpad
+
+    def _encode_dsa_der(cls, r, s):
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:550]: cls needed?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+
+    def _bytes_to_long(cls, b):
+        if hasattr(int, 'from_bytes'):
+            return int.from_bytes(b, byteorder='big', signed=False)
+        else:
+            return long(codecs.encode(b, encoding='hex'), 16)
+    self._bytes_to_long = _bytes_to_long
+    _bytes_to_long = _bytes_to_long
+
+    def _block_pad_unpad(cls, siz, bs, go):
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:544]: cls needed?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+    convert_public_key = convert_public_key
+
+    def convert_private_key(cls, packet):
+        return cls.convert_key(packet, True)
+    self.convert_private_key = convert_private_key
+    convert_private_key = convert_private_key
+
+    def _bytes_to_long(cls, b):
+        if hasattr(int, 'from_bytes'):
+            return int.from_bytes(b, byteorder='big', signed=False)
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:538]: there is a "long_to_bytes" function and a "bytes_from_long" in Crypto/Util/Number.star that can be used here I think?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 11]
```diff
@@ -0,0 +1,585 @@
...
+        else:
+            return long(codecs.encode(b, encoding='hex'), 16)
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:540]: I don't think this works. See above comment

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 12]
```diff
@@ -0,0 +1,585 @@
...
+    self._bytes_to_long = _bytes_to_long
+    _bytes_to_long = _bytes_to_long
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:542]: Needed?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+        return cls.convert_key(packet, False)
+    self.convert_public_key = convert_public_key
+    convert_public_key = convert_public_key
+
+    def convert_private_key(cls, packet):
+        return cls.convert_key(packet, True)
+    self.convert_private_key = convert_private_key
+    convert_private_key = convert_private_key
+
+    def _bytes_to_long(cls, b):
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:536]: Do we need cls?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 14]
```diff
@@ -0,0 +1,585 @@
...
+
+    def convert_public_key(cls, packet):
+        return cls.convert_key(packet, False)
+    self.convert_public_key = convert_public_key
+    convert_public_key = convert_public_key
+
+    def convert_private_key(cls, packet):
+        return cls.convert_key(packet, True)
+    self.convert_private_key = convert_private_key
+    convert_private_key = convert_private_key
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:534]: Needed?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 15]
```diff
@@ -0,0 +1,585 @@
...
+                return rsa.RSAPrivateNumbers(p, q, d, dmp1, dmq1, u, public).private_key(default_backend())
+            else:
+                return public.public_key(default_backend())
+    self.convert_key = convert_key
+    convert_key = convert_key
+
+    def convert_public_key(cls, packet):
+        return cls.convert_key(packet, False)
+    self.convert_public_key = convert_public_key
+    convert_public_key = convert_public_key
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:529]: Needed?

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star [hunk 16]
```diff
@@ -0,0 +1,585 @@
...
+                p = cls._bytes_to_long(packet.key['q'])
+                q = cls._bytes_to_long(packet.key['p'])
+                dmp1 = rsa.rsa_crt_dmp1(d, p)
+                dmq1 = rsa.rsa_crt_dmp1(d, q)
+                u = cls._bytes_to_long(packet.key['u'])
+                return rsa.RSAPrivateNumbers(p, q, d, dmp1, dmq1, u, public).private_key(default_backend())
+            else:
+                return public.public_key(default_backend())
+    self.convert_key = convert_key
+    convert_key = convert_key
```
Comment (mahmoudimus) [inline larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:524]: Needed?

**Review by mahmoudimus — CHANGES_REQUESTED** [review]

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star (hunk 2 above)
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:102]: It's here https://github.com/singpolyma/OpenPGP-Python/blob/master/OpenPGP/__init__.py#L1637 
gonna change it to something like Packet_tags...

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star (hunk 1 above)
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:6]: got it, thanks!

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star (hunk 3 above)
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:117]: That's so cool! will update it

File: larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star (hunk 4 above)
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/OpenPGP/__init__.star:232]: It was not finished, updated.

File: larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star (hunk 12 above)
Comment (la-luo) [reply larky/src/main/resources/vendor/Crypto/OpenPGP/cryptography.star:542]: We probably don't need this library at all. It was converted by py2star and I haven't start working on it

**Pull request closed without merging**